source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+
+  ir::Function* calleeFn = id2function_[call_inst_itr->GetSingleWordOperand(
+      kSpvFunctionCallFunctionId)];
+
+  // Map parameters to actual arguments.
+  MapParams(calleeFn, call_inst_itr, &callee2caller);
+
+  // Define caller local variables for all callee variables and create map to
+  // them.
+  CloneAndMapLocals(calleeFn, new_vars, &callee2caller);
+
+  // Create return var if needed.
+  uint32_t returnVarId = CreateReturnVar(calleeFn, new_vars);
+
+  // Clone and map callee code. Copy caller block code to beginning of
+  // first block and end of last block.
+  bool prevInstWasReturn = false;
+  uint32_t returnLabelId = 0;
+  bool multiBlocks = false;
+  const uint32_t calleeTypeId = calleeFn->type_id();
+  std::unique_ptr<ir::BasicBlock> new_blk_ptr;
+  calleeFn->ForEachInst([&new_blocks, &callee2caller, &call_block_itr,
+                         &call_inst_itr, &new_blk_ptr, &prevInstWasReturn,
+                         &returnLabelId, &returnVarId, &calleeTypeId,
+                         &multiBlocks, &postCallSB, &preCallSB, this](
+      const ir::Instruction* cpi) {
+    switch (cpi->opcode()) {
+      case SpvOpFunction:
+      case SpvOpFunctionParameter:
+      case SpvOpVariable:
+        // Already processed
+        break;
+      case SpvOpLabel: {
+        // If previous instruction was early return, insert branch
+        // instruction to return block.
+        if (prevInstWasReturn) {
+          if (returnLabelId == 0) returnLabelId = this->TakeNextId();
+          AddBranch(returnLabelId, &new_blk_ptr);
+          prevInstWasReturn = false;
+        }
+        // Finish current block (if it exists) and get label for next block.
+        uint32_t labelId;
+        bool firstBlock = false;
+        if (new_blk_ptr != nullptr) {
+          new_blocks->push_back(std::move(new_blk_ptr));
+          // If result id is already mapped, use it, otherwise get a new
+          // one.
+          const uint32_t rid = cpi->result_id();
+          const auto mapItr = callee2caller.find(rid);
+          labelId = (mapItr != callee2caller.end()) ? mapItr->second
+                                                    : this->TakeNextId();
+        } else {
+          // First block needs to use label of original block
+          // but map callee label in case of phi reference.
+          labelId = call_block_itr->label_id();
+          callee2caller[cpi->result_id()] = labelId;
+          firstBlock = true;
+        }
+        // Create first/next block.
+        new_blk_ptr.reset(new ir::BasicBlock(NewLabel(labelId)));
+        if (firstBlock) {
+          // Copy contents of original caller block up to call instruction.
+          for (auto cii = call_block_itr->begin(); cii != call_inst_itr;
+               cii++) {
+            std::unique_ptr<ir::Instruction> cp_inst(new ir::Instruction(*cii));
+            // Remember same-block ops for possible regeneration.
+            if (cp_inst->opcode() == SpvOpSampledImage) {
```

Comment:
Reiterating the need to handle OpImage similarly.  More details:  https://github.com/KhronosGroup/SPIRV-Tools/pull/534#discussion_r108220528